src/serializer.cpp

```diff
@@ -144,6 +161,180 @@ prepare() ->
             error::expect_100_continue);
     }
 
+    if( is_compressed_ )
+    {
+        auto on_end = [&]
+        {
+            std::size_t n = 0;
+            if(out_.data() == hp_)
+                ++n;
+
+            for(buffers::const_buffer const& b : tmp0_.data())
+                out_[n++] = b;
+
+            auto cbs = const_buffers_type(
+                out_.data(), out_.size());
+
+            BOOST_ASSERT(buffers::buffer_size(cbs) > 0);
+
+            return cbs;
+        };
+
+        if(! zlib_filter_ )
+        {
+            auto& svc =
+                ctx_->get_service<
+                    zlib::deflate_decoder_service>();
+
+            zlib_filter_ = &svc.make_filter(filter_ws_);
+        }
+
+        auto& zbuf = tmp1_;
+
+        if( st_ == style::source )
+        {
+            auto results = src_->read(
+                zbuf.prepare(zbuf.capacity()));
+            more_ = !results.finished;
+            zbuf.commit(results.bytes);
+        }
+
+        buffers::mutable_buffer chunk_header;
+        if( is_chunked_ )
+        {
+            auto mbs = tmp0_.prepare(detail::chunk_header_len_);
+            auto buf = *mbs.begin();
+            if( buf.size() == 0 )
+            {
+                auto p = mbs.begin();
+                ++p;
+                buf = *p;
+            }
+
+            if( buf.size() < detail::chunk_header_len_ )
+                detail::throw_length_error();
+
+            std::memset(
+                buf.data(), 0x00, detail::chunk_header_len_);
+            tmp0_.commit(detail::chunk_header_len_);
+
+            chunk_header = buf;
+        }
+
+        auto get_output = [&]() -> buffers::mutable_buffer
```

Comment:
I'm not really a fan of these lambda expressions because they make the code harder to read